### ISSUE TITLE
Minimum and maximum values of architectural parameters

### ISSUE THREAD

**Issue by Bebra777228** [nomadkaraoke/python-audio-separator#138]
In the demucs_params, the segment_size is set to 'Default'. I would like to understand what this means and what specific value 'Default' implies.

Additionally, it would be helpful to know at least the approximate minimum and maximum values for each parameter across different architectures.

---
_I have already written this text as part of another issue, but since there has been no response to it for several days, I decided to duplicate it in a new issue and close the old one =)_

**Comment by beveradb** [comment]
I'd like to take a moment to encourage you (and anyone else) to _read the code_ and figure out the answer yourself when you have a question like this rather than just raising an issue 😄 

Even better, if you want to contribute back to the community, you can raise a PR improving the documentation or write up a post explaining your findings once you've dug into the code to figure it out!

The mental load of maintaining this repository for free is already more than I would like, which is why you get very slow responses from me and why I would prefer to help folks help _themselves_ rather than expecting me to answer.

If you try your best to figure it out but get stuck / struggle, I'm much happier to help _teach_ you how things work / how to help yourself, but it would be helpful if you could do the initial attempt and write up your findings / explain where you get stuck and then ask me for help pointing you in the right direction to unstick you! 

Anyway, for this specific case you asked about (demucs), I'll answer the question for you and show you my approach:

------------------------------------

I started out by opening the repo in vscode and searching for the string `Default` across the whole codebase.
That gave only a few results, giving us a few good starts for where to look:

<img width="335" alt="image" src="https://github.com/user-attachments/assets/c1f72a2d-9530-43e2-9cbf-8838ad26b68f">

- Starting in the top level Separator class `audio_separator/separator/separator.py`, on [this line](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/separator.py#L96C9-L96C22), we see: `demucs_params={"segment_size": "Default"`
- Further [down the file](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/separator.py#L170C9-L170C34), we can see that the `demucs_params` value gets passed in under a `Demucs` key in `self.arch_specific_params`
- `arch_specific_params` then gets [passed in](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/separator.py#L734) to the instance of whatever separator architecture class is loaded, which we can see from [this line](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/separator.py#L719) is `demucs_separator.DemucsSeparator`
- So, I then open up `audio_separator/separator/architectures/demucs_separator.py` and we can see from [this line](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/architectures/demucs_separator.py#L47) that the `segment_size` value just gets passed through to be set as `self.segment_size` in this class instance.
- However, we also see a comment which has some useful detail on what values may be valid for this parameter:
```
        # Adjust segments to manage RAM or V-RAM usage:
        # - Smaller sizes consume less resources.
        # - Bigger sizes consume more resources, but may provide better results.
        # - "Default" picks the optimal size.
        # DEMUCS_SEGMENTS = (DEF_OPT, '1', '5', '10', '15', '20',
        #           '25', '30', '35', '40', '45', '50',
        #           '55', '60', '65', '70', '75', '80',
        #           '85', '90', '95', '100')
```
- Searching for `segment_size` now in this file, we can see [further down](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/architectures/demucs_separator.py#L105) that the value is passed into a function called `demucs_segments`, which is [imported](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/architectures/demucs_separator.py#L7) from `uvr_lib_v5.demucs.apply`
- So we open up `audio_separator/separator/uvr_lib_v5/demucs/apply.py`, where we can see the [demucs_segments function](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/uvr_lib_v5/demucs/apply.py#L264)
- This function is pretty badly written to be honest (not my code) but it looks like there are three main cases:
  a. If demucs_segment is "Default": **Set segment to None**
  b. If demucs_segment is a valid integer: **Set segment to that integer value**
  c. If demucs_segment is neither "Default" nor a valid integer: **Set segment to None (same as "Default" case)**
- After determining the segment value, it attempts to apply this value to the model(s):
  - If demucs_model is a BagOfModels, it tries to set the segment attribute for each sub-model.
  - If demucs_model is not a BagOfModels, it tries to set the segment attribute directly.
  - Finally, it returns the (potentially modified) demucs_model.
- So, if this function just sets the value to `None` when it's set to `Default`, what actually is the "default" value?
- Well, if we go back to the [demucs_separator](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/architectures/demucs_separator.py#L103) code, the object being modified by that `demucs_segments` function is an instance of the `HDemucs` class from `audio_separator/separator/uvr_lib_v5/demucs/hdemucs.py`, so I open that up, and...
- Bingo, this class has a `segment` property which is instantiated with a [default value](https://github.com/nomadkaraoke/python-audio-separator/blob/main/audio_separator/separator/uvr_lib_v5/demucs/hdemucs.py#L407) of `40`

So, in short, the default value is 40.
Hope this process helps you understand how you can answer questions like this yourself in future 😄

**Comment by Bebra777228** [comment]
Oh yes, I have already resolved this issue. While waiting for a response, I completely forgot about it.

The issue arose due to an error that occurred when changing the segment_size parameter in demucs_params. The error was related to a float. I fixed it by reverting to the Default value.
